src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,162 @@
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value);
+        if (head == null) {
+            head = tail = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexForAdd(index);
+        Node<T> newNode = new Node<>(value);
+        if (head == null) {
+            head = tail = newNode;
+        } else if (index == 0) {
+            newNode.next = head;
+            head.next.prev = newNode;
+            head = newNode;
+        } else if (index == size) {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        } else {
+            Node<T> current = takeElementByIndex(index);
+            newNode.next = current;
+            current.prev.next = newNode;
+            newNode.prev = current.prev;
+            current.prev = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T values : list) {
+            add(values);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> elementByIndex = takeElementByIndex(index);
+        return (elementByIndex == null) ? null : elementByIndex.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> elementByIndex = takeElementByIndex(index);
+        if (elementByIndex == null) {
+            return null;
+        }
+        T valueToReturn = elementByIndex.item;
+        elementByIndex.item = value;
+        return valueToReturn;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> elementByIndex = removeElement(index);
+        return elementByIndex.item;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        for (int index = 0; index < size; index++) {
+            if (current == null) {
+                return false;
+            } else if (Objects.equals(object, current.item)) {
+                removeElement(index);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return (head == null);
+    }
+
+    private Node<T> takeElementByIndex(int index) {
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            if (current == null) {
+                return null;
+            } else {
+                current = current.next;
+            }
+        }
+        return current;
+    }
+
+    private Node<T> removeElement(int index) {
```

Comment:
```suggestion
    private Node<T> unlink(Node<T> node) {
```
you should have unlink method

![image](https://user-images.githubusercontent.com/81781479/198707253-5531521c-8a1b-4c6b-8c21-755eec00a872.png)